// @agent-tars/core/dist/browser/tools/navigation.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { Tool, z } from "@mcp-agent/core";
function createNavigationTools(logger, browserGUIAgent) {
    const navigateTool = new Tool({
        id: 'browser_navigate',
        description: '[browser] Navigate to a URL',
        parameters: z.object({
            url: z.string().describe('URL to navigate to')
        }),
        function: async ({ url })=>{
            try {
                if (!browserGUIAgent) return {
                    status: 'error',
                    message: 'GUI Agent not initialized'
                };
                const page = await browserGUIAgent.getPage();
                await page.goto(url);
                return {
                    status: 'success',
                    url,
                    message: "Navigated success"
                };
            } catch (error) {
                logger.error(`Error navigating to URL: ${error}`);
                return {
                    status: 'error',
                    url,
                    message: `Failed to navigate: ${error instanceof Error ? error.message : String(error)}`
                };
            }
        }
    });
    const backTool = new Tool({
        id: 'browser_go_back',
        description: '[browser] Go back to the previous page, or close tab if no history exists',
        parameters: z.object({}),
        function: async ()=>{
            try {
                if (!browserGUIAgent) return {
                    status: 'error',
                    message: 'GUI Agent not initialized'
                };
                const page = await browserGUIAgent.getPage();
                const hasHistory = await page.evaluate(()=>window.history.length > 1);
                if (hasHistory) {
                    await page.goBack();
                    return {
                        status: 'success',
                        message: 'Navigated back'
                    };
                }
                {
                    const browser = page.browser();
                    const pages = await browser.pages();
                    if (pages.length > 1) {
                        const currentUrl = page.url();
                        await page.close();
                        logger.info(`Closed tab with URL ${currentUrl} (no history to go back to)`);
                        return {
                            status: 'success',
                            message: 'No history available, closed the current tab instead',
                            action: 'closed'
                        };
                    }
                    await page.goto('about:blank');
                    return {
                        status: 'success',
                        message: 'No history available and only one tab open, navigated to blank page',
                        action: 'blank'
                    };
                }
            } catch (error) {
                logger.error(`Error handling back navigation: ${error}`);
                return {
                    status: 'error',
                    message: `Failed to navigate back: ${error instanceof Error ? error.message : String(error)}`
                };
            }
        }
    });
    const forwardTool = new Tool({
        id: 'browser_go_forward',
        description: '[browser] Go forward to the next page',
        parameters: z.object({}),
        function: async ()=>{
            try {
                if (!browserGUIAgent) return {
                    status: 'error',
                    message: 'GUI Agent not initialized'
                };
                const page = await browserGUIAgent.getPage();
                await page.goForward();
                return {
                    status: 'success',
                    message: 'Navigated forward'
                };
            } catch (error) {
                logger.error(`Error navigating forward: ${error}`);
                return {
                    status: 'error',
                    message: `Failed to navigate forward: ${error instanceof Error ? error.message : String(error)}`
                };
            }
        }
    });
    const refreshTool = new Tool({
        id: 'browser_refresh',
        description: '[browser] Refresh the current page',
        parameters: z.object({}),
        function: async ()=>{
            try {
                if (!browserGUIAgent) return {
                    status: 'error',
                    message: 'GUI Agent not initialized'
                };
                const page = await browserGUIAgent.getPage();
                await page.reload();
                return {
                    status: 'success',
                    message: 'Page refreshed'
                };
            } catch (error) {
                logger.error(`Error refreshing page: ${error}`);
                return {
                    status: 'error',
                    message: `Failed to refresh: ${error instanceof Error ? error.message : String(error)}`
                };
            }
        }
    });
    return [
        navigateTool,
        backTool,
        forwardTool,
        refreshTool
    ];
}
export { createNavigationTools };
