// @agent-tars/server/dist/api/controllers/system.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
function healthCheck(req, res) {
    res.status(200).json({
        status: 'ok'
    });
}
function getModelInfo(req, res) {
    try {
        var _server_appConfig_model, _server_appConfig, _server_appConfig_model1, _server_appConfig1;
        const server = req.app.locals.server;
        const modelInfo = {
            provider: process.env.MODEL_PROVIDER || (null == (_server_appConfig = server.appConfig) ? void 0 : null == (_server_appConfig_model = _server_appConfig.model) ? void 0 : _server_appConfig_model.provider) || 'Default Provider',
            model: process.env.MODEL_NAME || (null == (_server_appConfig1 = server.appConfig) ? void 0 : null == (_server_appConfig_model1 = _server_appConfig1.model) ? void 0 : _server_appConfig_model1.id) || 'Default Model'
        };
        res.status(200).json(modelInfo);
    } catch (error) {
        console.error('Error getting model info:', error);
        res.status(500).json({
            error: 'Failed to get model information'
        });
    }
}
export { getModelInfo, healthCheck };
