// @agent-tars/core/dist/browser-manager.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import * as __WEBPACK_EXTERNAL_MODULE__agent_infra_browser_45fd5ecc__ from "@agent-infra/browser";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class BrowserManager {
    static getInstance(logger) {
        if (!BrowserManager.instance) BrowserManager.instance = new BrowserManager(logger);
        return BrowserManager.instance;
    }
    getBrowser() {
        if (!this.browser) {
            this.logger.info('Creating browser instance (not launched yet)');
            this.browser = new __WEBPACK_EXTERNAL_MODULE__agent_infra_browser_45fd5ecc__.LocalBrowser({
                logger: this.logger.spawn('LocalBrowser')
            });
        }
        return this.browser;
    }
    async launchBrowser(options = {}) {
        if (this.isLaunched) {
            this.logger.info('Browser already launched, skipping launch');
            return;
        }
        try {
            this.logger.info('🌐 Launching browser instance...');
            const browser = this.getBrowser();
            await browser.launch(options);
            this.isLaunched = true;
            this.logger.success('✅ Browser instance launched successfully');
        } catch (error) {
            this.logger.error(`❌ Failed to launch browser: ${error}`);
            throw error;
        }
    }
    isLaunchingComplete() {
        return this.isLaunched;
    }
    async isBrowserAlive() {
        if (!this.browser || !this.isLaunched) return false;
        return this.browser.isBrowserAlive();
    }
    async closeBrowser() {
        if (this.browser && this.isLaunched) try {
            this.logger.info('Closing browser instance...');
            await this.browser.close();
            this.logger.info('Browser instance closed successfully');
        } catch (error) {
            this.logger.error(`Error closing browser: ${error}`);
        } finally{
            this.isLaunched = false;
        }
    }
    static resetInstance() {
        BrowserManager.instance = null;
    }
    constructor(logger){
        _define_property(this, "browser", null);
        _define_property(this, "isLaunched", false);
        _define_property(this, "logger", void 0);
        this.logger = logger.spawn('BrowserManager');
        this.logger.info('Browser manager initialized (browser not launched yet)');
    }
}
_define_property(BrowserManager, "instance", null);
export { BrowserManager };
